pep-9999.rst

```diff
@@ -0,0 +1,874 @@
+
+These considerations have guided the design and updates proposed in this PEP.
+
+Reusing the existing `license` field with license expressions will provide a an
+intuitive and more structured way to express the license of a distribution using
+a well defined syntax and well known license ids.
+
+Over time, recommending the usage of these expressions will help Python package
+publishers improve the clarity of their license documentation to the benefit of
+packages authors, consumers and redistributors.
+
+
+Core Metadata Specification updates
+===================================
+
+The canonical source for the names and semantics of each of the supported
+metadata fields is the Core Metadata Specification [#cms]_ document.
+
+The details of the updates considered to the Core Metadata Specification [#cms]_
+document as part of his PEP are detailed here and will be added to the
+canonical source once this PEP is approved.
+
+
+Added in Version 2.2
+--------------------
+
+License-File (multiple use)
+:::::::::::::::::::::::::::
+
+The is a string that is a package-root relative path to a license file. The
+license file content __must__ be UTF-8-encoded text.
+
+Build tools should honor this field and include the corresponding license
+file(s) in the built package.
+
+Note that this is field already exists in `wheel` and `setuptools` with this
+behaviour. Therefore this addition to the Core metadata is only recognizing and
+documenting an existing practice as used in `wheels` (with the `license_file`
+and `license_files` `setup.cfg` `[metadata]` entries) and in `setuptools`
+`license_file` `setup()` argument.
+
+
+Changed in Version 2.2
+----------------------
+
+License (optional)
+::::::::::::::::::
+
+Text indicating the license covering the distribution. This text can be either a
+valid `License Expression` as defined here or any free text.
+
+Publishing tools SHOULD issue an informational warning if this field is empty or
+missing or is not a valid `License Expression` as defined here. Build tools MAY
+issue such a warning too.
+
+
+License Expression syntax
+'''''''''''''''''''''''''
+
+A `License Expression` is a string using the SPDX license expression syntax as
+documented in the SPDX specification [#spdx]_ using either Version 2.1
+[#spdx21]_ or a later compatible version. SPDX is a working group at the Linux
+Foundation that defines a standard way to exchange package information.
+
+When used in the `License` field and as a specialization of the SPDX license
+expression definition, a `License Expression` can use the following license
+identifiers:
+
+- any SPDX-listed license short-form identifiers that are published in the 
+  SPDX License List [#spdxlist]_ using either Version 3.6 of this list or any
+  later compatible version. Note that the SPDX working group never removes any
+  license identifiers: instead they may only one as obsolete.
+
+- the `Public-Domain` and `Proprietary` strings to support generic ids that are
```

Comment:
I understand and agree with including the `Proprietary` option here, but what's the value in allowing `Public-Domain` over encouraging folks to instead write the more explicit and legally portable `CC0-1.0`?

PyPI is a global database, and the meaning associated with the term "public domain" is thoroughly dependent on the specific legal jurisdiction involved.

Reply:
> I understand and agree with including the `Proprietary` option here, but what's the value in allowing `Public-Domain` over encouraging folks to instead write the more explicit and legally portable `CC0-1.0`?
> 
> PyPI is a global database, and the meaning associated with the term "public domain" is thoroughly dependent on the specific legal jurisdiction involved.

I am trying to adapt to the existing Classifiers as much as possible, and here is where we do not have a direct SPDX license id for a Classifier:
- `License :: DFSG approved` and `License :: OSI Approved` : these are most informational than a license statement. IMHO they do not need to be carried over

- `License :: Free For Educational Use`, `License :: Free For Home Use`, `License :: Free for non-commercial use`, `License :: Freely Distributable`, `License :: Free To Use But Restricted`, `License :: Freeware`, `License :: Other/Proprietary License`. The would all become `Proprietary`.

- `License :: GUST Font License 1.0`, `License :: GUST Font License 2006-09-30` we could have them added @ SPDX but no package uses them in PyPI today.

- `License :: Public Domain`  is not in SPDX (`CC0-1.0` is) and even if I would prefer having a single id there, a US Gvt. (or US) Public domain dedication is not in SPDX either, so we could keep `Public-Domain` for these cases. We have ~ 900 of these https://pypi.org/search/?c=License+%3A%3A+Public+Domain ... I like to be prescriptive but public domain dedication are the simplest form of licensing and while we could encourage folks using a bare "Public-Domain" to switch to `CC0-1.0` or the `Unlicense`, it may be best to keep something for that?
FWIW, even the PEPs license is a choice of Public Domain or CC0-1.0.

So I am fine with dropping a bare Public Domain, yet I would thread lightly (though it would surely rejoice our German friends and other countries that have no concept of "Public Domain")

Reply:
Aye, I wrote my original comment before seeing that the rationale related to being able to drop the use of the corresponding license classifier while providing a clear migration path.

Perhaps we could encourage the use of the more global `Public-Domain OR CC0-1.0` as the replacement for `License :: Public Domain` (akin to the updated PEP licensing) while still making it clear that the bare `Public-Domain` expression is acceptable at a tooling level?

(Note: DFSG approved and OSI Approved aren't intended to be used as classifiers in their own right, they're just a way of grouping the other classifiers, so I agree they don't need a directly translation)

Reply:
I just pushed with an update https://github.com/pombredanne/spdx-pypi-pep/pull/2/commits/a61294ef6653c5a36bab94590372e0b1427f7c2d before seeing this....

Actually `Public-Domain OR CC0-1.0` is a great example for using an expression!
So what we could do could be this:

- add `Public-Domain` as a license id
- in the tools-related section state that tools SHOULD display a warning when a bare `Public-Domain` is used and and suggest the use of the more global `Public-Domain OR CC0-1.0` expression?

Reply:
@VanL would this make sense?

Reply:
As I commented else where, don't extend license list, it would destroy the utility of using commonly defined syntax.

Use `LicenseRef-PublicDomain` if you really need.

It's ugly, but it also makes you think why not use `CC0`.

Reply:
> Use LicenseRef-PublicDomain if you really need.

The problem is that this `LicenseRef-PublicDomain` is not a valid license id when used outside of an SPDX document (which is the case here). So using  `Public-Domain` which is not listed is no more nor less valid than `LicenseRef-PublicDomain`.

There is a pending proposal and work on creating namespaced license ids but that is not yet fully firmed up and this could provide a way. 

For similar uses, `npm` adopted `UNLICENSED` for proprietary licenses which is not great IMHO.
And we use `OTHER`` for anything not SPDX-listed at https://clearlydefined.io/ as a wart while namespaces are worked out.

Reply:
@pombredanne I try to read SPDX specification, and it's very unclear. The appendix IV doesn't hint that in some places where "SPDX License Expression" is used `license-ref`s aren't allowed.

As concrete example, *appendix V: Using SPDX short identifiers in Source Files*

> The SPDX License Identifier syntax may consist of  a single license (represented by a short identifier from
the SPDX license list) or a compound set of licenses (represented by joining together multiple licenses using
the license expression syntax).

That doesn't exclude licence references explicitly.

Also following section

> Representing Multiple Licenses
> Multiple licenses can be represented using a SPDX license expression as defined in Appendix IV.  

doesn't restrict the syntax.

---

So, as far as I can  read the SPDX specification, `LicenseRef-PublicDomain` indeed isn't valid *License identifier* as it's not in SPDX License List, but it's valid *SPDX Expression*.

The `SPDX-License-Identifier` in source files implies it could be just an identifier, but someone else already noted that it's badly named (but maybe we are stuck with it due legacy reasons): https://github.com/spdx/spdx-spec/issues/82

---

In Cabal packages, the license file is soft-mandatory so `LicenseRef-YourCompanyNameLicense` can be understood as a reference to that file.

---

Point I try to make: is that if someone is assembling an actual SPDX document for a such "Proprietary" package, then they will eventually write `LicenseRef-Something` for this proprietary licenses, won't they?

Reply:
https://spdx.org/spdx-specification-21-web-version#h.1v1yuxt seems to be the main authoritative section on the use of the `LicenseRef-` prefix in license names, and I agree with @phadej that it would make sense to structure our extensions such that they can be composed into a valid SPDX document simply by providing an `ExtractedLicensingInfo` section that references the published PEP and the corresponding `LicenseID` entry.

If the PEP doesn't handle that, then anyone trying to compose SPDX documents based on Python package metadata expressions would need to come up with their own way of mapping it. By contrast, if the PEP defines a suggested mapping up front, then folks would only need to define their own if they were analysing a project that used `LicenseRef-PublicDomain` and `LicenseRef-Proprietry` for multiple purposes.

Reply:
@ncoghlan @phadej I am siding with you and I am updating the draft accordingly! Using a `LicenseRef-` prefix is going to be cleaner in the end even if not in the context of a full SPDX document so `LicenseRef-PublicDomain` and `LicenseRef-Proprietary` it will be together with a link to the SPDX spec. :+1:

Reply:
This has been added in the PEP draft with a61294ef6653c5a36bab94590372e0b1427f7c2d